ScriptFunctionCall::call() can return an empty JSValue if the watchdog timer fires, callers should check for this

yury-s opened the pull request:
#### a4eed62b176cca45404f2c11f8a3dd2c4d3fe5df
<pre>
ScriptFunctionCall::call() can return an empty JSValue if the watchdog timer fires, callers should check for this
<a href="https://bugs.webkit.org/show_bug.cgi?id=165875">https://bugs.webkit.org/show_bug.cgi?id=165875</a>

Reviewed by Devin Rousso.

ScriptFunctionCall::call() may return empty JSValue from several places,
the callers now check for emptiness first before accessing the value.

Unfortunately, I don&apos;t have a reliable repro which could be converted
to a layout test like the one in <a href="https://github.com/WebKit/WebKit/commit/11d211bca821fa4803d6da95c857a04b7f32c46a">https://github.com/WebKit/WebKit/commit/11d211bca821fa4803d6da95c857a04b7f32c46a</a>
even though the symptoms are similar.

* Source/JavaScriptCore/inspector/InjectedScript.cpp:
(Inspector::InjectedScript::wrapObject const):
(Inspector::InjectedScript::wrapJSONString const):
(Inspector::InjectedScript::wrapTable const):
(Inspector::InjectedScript::previewValue const):
(Inspector::InjectedScript::createCommandLineAPIObject const):

Canonical link: <a href="https://commits.webkit.org/270739@main">https://commits.webkit.org/270739@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56026284c9f9d857e98c1eb7db69368609f05591

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26006 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4618 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27289 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28106 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23821 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6381 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2047 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23889 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3498 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28686 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3118 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23360 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29421 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22657 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23734 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23737 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27307 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25226 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1360 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32671 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4540 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7089 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6313 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3608 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso reviewed (commented):
why is it just these few spots and not all the other places where `callFunctionWithEvalEnabled` is called (not including the callsites that already have this change/behavior)?

e.g. `InjectedScript::findObjectById`, `InjectedScriptBase::makeAsyncCall`, `InjectedScriptModule::ensureInjected`, etc.

NIT: typo in commit message `Unfortunatelya`

yury-s committed:
ScriptFunctionCall::call() can return an empty JSValue if the watchdog timer fires, callers should check for this
https://bugs.webkit.org/show_bug.cgi?id=165875

Reviewed by Devin Rousso.

ScriptFunctionCall::call() may return empty JSValue from several places,
the callers now check for emptiness first before accessing the value.

Unfortunately, I don't have a reliable repro which could be converted
to a layout test like the one in https://github.com/WebKit/WebKit/commit/11d211bca821fa4803d6da95c857a04b7f32c46a
even though the symptoms are similar.

* Source/JavaScriptCore/inspector/InjectedScript.cpp:
(Inspector::InjectedScript::wrapObject const):
(Inspector::InjectedScript::wrapJSONString const):
(Inspector::InjectedScript::wrapTable const):
(Inspector::InjectedScript::previewValue const):
(Inspector::InjectedScript::createCommandLineAPIObject const):

Canonical link: https://commits.webkit.org/270739@main
